core/src/Kokkos_MathematicalFunctions.hpp

```diff
@@ -0,0 +1,232 @@
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_MATHEMATICAL_FUNCTIONS_HPP
+#define KOKKOS_MATHEMATICAL_FUNCTIONS_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <cmath>
+#include <algorithm>
+#include <type_traits>
+
+#ifdef KOKKOS_ENABLE_SYCL
+#include <CL/sycl.hpp>
+#endif
+
+namespace Kokkos {
+namespace Experimental {
+
+#if defined(KOKKOS_ENABLE_SYCL)
+#define NAMESPACE_MATH_FUNCTIONS cl::sycl
+#else
+#define NAMESPACE_MATH_FUNCTIONS std
+#endif
+
+#define KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, RETURNTYPE, ARGTYPE) \
+  KOKKOS_INLINE_FUNCTION RETURNTYPE FUNC(ARGTYPE x) {                        \
+    using NAMESPACE_MATH_FUNCTIONS::FUNC;                                    \
+    return FUNC(x);                                                          \
+  }
+
+#define KOKKOS_IMPL_UNARY_FUNCTION_INTEGRAL(FUNC, RETURNTYPE)              \
+  template <typename Integer,                                              \
+            typename = std::enable_if_t<std::is_integral<Integer>::value>> \
+  KOKKOS_INLINE_FUNCTION RETURNTYPE FUNC(Integer x) {                      \
+    return Kokkos::Experimental::FUNC(static_cast<double>(x));             \
+  }
+
+#define KOKKOS_IMPL_BINARY_FUNCTION_FLOATING_POINT(FUNC, TYPE) \
+  KOKKOS_INLINE_FUNCTION TYPE FUNC(TYPE x, TYPE y) {           \
+    using NAMESPACE_MATH_FUNCTIONS::FUNC;                      \
+    return FUNC(x, y);                                         \
+  }
+
+// NOTE long double overloads are not available on the device
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) || \
+    defined(KOKKOS_ENABLE_SYCL) || defined(KOKKOS_ENABLE_OPENMPTARGET)
+
+#define KOKKOS_IMPL_BINARY_FUNCTION_ARITHMETIC(FUNC)                         \
+  template <typename Arithmetic1, typename Arithmetic2,                      \
+            typename = std::enable_if_t<                                     \
+                std::is_arithmetic<Arithmetic1>::value &&                    \
+                std::is_arithmetic<Arithmetic2>::value &&                    \
+                !std::is_same<Arithmetic1, long double>::value &&            \
+                !std::is_same<Arithmetic2, long double>::value>>             \
+  KOKKOS_INLINE_FUNCTION double FUNC(Arithmetic1 x, Arithmetic2 y) {         \
```

Comment:
Should this always return `double`? This will implicitly cast the return value from `FUNC` to `double` and I'm not sure we want that. For example, what if `FUNC` returns a quad precision number?

Reply:
Yes, if non of the two argument can be `long double`, then the return type is always `double`.
See https://eel.is/c++draft/c.math#cmath.syn-2

About quadruple-precision fp, as far as I know there is no native support for floating-point types with more than double precision.